src/Orleans/Configuration/ProviderConfiguration.cs

```diff
@@ -193,23 +193,39 @@ public IList<IProvider> Children
     [Serializable]
     public class ProviderCategoryConfiguration
     {
-        public string Name { get; set; }
+        public ProviderCategoryConfiguration()
+        {
+            Providers = new Dictionary<string, IProviderConfiguration>();
+        }
 
-        public IDictionary<string, IProviderConfiguration> Providers { get; set; }
+        internal ProviderCategoryConfiguration(string name) : this()
```

Comment:
Performance. Additional noise not adding anything. I would personally be confused seeing the code which does such checks inside internal code. The first question I would ask: why this is not done on the boundary?

But you maybe right, that even better option would to make this class completely internal. I have no idea why would end-user need to create custom provider category at all.